File: js/src/result.ts
/**
 * Inference result types for BayesianGPU
 *
 * This module defines the types returned by the inference engine.
 *
 * @module
 */

/**
 * Inference configuration used for sampling
 */
export interface InferenceConfig {
  /** Number of samples to draw (after warmup) */
  numSamples?: number;
  /** Number of warmup iterations */
  numWarmup?: number;
  /** Number of parallel chains */
  numChains?: number;
  /** Target acceptance probability (0-1) */
  targetAccept?: number;
  /** Random seed for reproducibility */
  seed?: number;
}

/**
 * MCMC diagnostics for a single parameter or overall
 */
export interface Diagnostics {
  /** R-hat convergence diagnostic (per parameter) */
  rhat: Record<string, number>;
  /** Effective sample size (per parameter) */
  ess: Record<string, number>;
  /** Number of divergent transitions */
  divergences: number;
}

/**
 * Configuration that was actually used (with defaults filled in)
 */
export interface ActualConfig {
  numSamples: number;
  numWarmup: number;
  numChains: number;
  stepSize: number;
}

/**
 * Complete inference result
 */
export interface InferenceResult {
  /** Samples for each parameter (flattened across chains) */
  samples: Record<string, number[]>;
  /** MCMC diagnostics */
  diagnostics: Diagnostics;
  /** Configuration used */
  config: ActualConfig;
}

/**
 * Summary statistics for a single parameter
 */
export interface ParameterSummary {
  /** Posterior mean */
  mean: number;
  /** Posterior standard deviation */
  std: number;
  /** 2.5th percentile (lower bound of 95% CI) */
  q025: number;
  /** 25th percentile */
  q25: number;
  /** Median (50th percentile) */
  q50: number;
  /** 75th percentile */
  q75: number;
  /** 97.5th percentile (upper bound of 95% CI) */
  q975: number;
  /** R-hat convergence diagnostic */
  rhat: number;
  /** Effective sample size */
  ess: number;
}

/**
 * Check if diagnostics indicate good convergence
 *
 * @param diagnostics - The diagnostics object
 * @returns true if all R-hat values are below 1.01
 */
export function isConverged(diagnostics: Diagnostics): boolean {
  return Object.values(diagnostics.rhat).every(
    (r) => r !== undefined && r < 1.01
  );
}

/**
 * Check if effective sample size is sufficient
 *
 * @param diagnostics - The diagnostics object
 * @param minEss - Minimum ESS required (default: 400)
 * @returns true if all ESS values are above the minimum
 */
export function hasSufficientESS(
  diagnostics: Diagnostics,
  minEss: number = 400
): boolean {
  return Object.values(diagnostics.ess).every(
    (e) => e !== undefined && e >= minEss
  );
}

/**
 * Get warning messages for any diagnostic issues
 *
 * @param result - The inference result
 * @returns Array of warning messages
 */
export function getWarnings(result: InferenceResult): string[] {
  const warnings: string[] = [];

  // Check R-hat values
  for (const [param, rhat] of Object.entries(result.diagnostics.rhat)) {
    if (rhat >= 1.1) {
      warnings.push(
        `${param}: R-hat = ${rhat.toFixed(3)} indicates poor convergence (should be < 1.01)`
      );
    } else if (rhat >= 1.01) {
      warnings.push(
        `${param}: R-hat = ${rhat.toFixed(3)} is marginal (ideally < 1.01)`
      );
    }
  }

  // Check ESS values
  for (const [param, ess] of Object.entries(result.diagnostics.ess)) {
    if (ess < 100) {
      warnings.push(
        `${param}: ESS = ${ess.toFixed(0)} is too low (should be > 100)`
      );
    } else if (ess < 400) {
      warnings.push(
        `${param}: ESS = ${ess.toFixed(0)} is low (ideally > 400)`
      );
    }
  }

  // Check divergences
  const totalSamples =
    result.config.numSamples * result.config.numChains;
  const divergenceRate = result.diagnostics.divergences / totalSamples;
  if (divergenceRate > 0.05) {
    warnings.push(
      `${result.diagnostics.divergences} divergences (${(divergenceRate * 100).toFixed(1)}%) - consider reparameterizing`
    );
  } else if (result.diagnostics.divergences > 0) {
    warnings.push(
      `${result.diagnostics.divergences} divergences detected`
    );
  }

  return warnings;
}

/**
 * Compute summary statistics for a parameter
 *
 * @param result - The inference result
 * @param paramName - Name of the parameter
 * @returns Summary statistics
 */
export function summarizeParameter(
  result: InferenceResult,
  paramName: string
): ParameterSummary {
  const samples = result.samples[paramName];
  if (!samples || samples.length === 0) {
    throw new Error(`Parameter '${paramName}' not found in results`);
  }

  const n = samples.length;

  // Compute mean
  const mean = samples.reduce((a, b) => a + b, 0) / n;

  // Compute standard deviation
  const variance =
    samples.reduce((sum, x) => sum + (x - mean) ** 2, 0) / (n - 1);
  const std = Math.sqrt(variance);

  // Sort for quantiles
  const sorted = [...samples].sort((a, b) => a - b);

  // Compute quantiles
  const quantile = (p: number): number => {
    const idx = p * (n - 1);
    const lower = Math.floor(idx);
    const upper = Math.min(lower + 1, n - 1);
    const weight = idx - lower;
    return sorted[lower] * (1 - weight) + sorted[upper] * weight;
  };

  return {
    mean,
    std,
    q025: quantile(0.025),
    q25: quantile(0.25),
    q50: quantile(0.5),
    q75: quantile(0.75),
    q975: quantile(0.975),
    rhat: result.diagnostics.rhat[paramName] ?? NaN,
    ess: result.diagnostics.ess[paramName] ?? NaN,
  };
}

/**
 * Compute summary statistics for all parameters
 *
 * @param result - The inference result
 * @returns Map of parameter names to their summaries
 */
export function summarizeAll(
  result: InferenceResult
): Record<string, ParameterSummary> {
  const summaries: Record<string, ParameterSummary> = {};
  for (const paramName of Object.keys(result.samples)) {
    summaries[paramName] = summarizeParameter(result, paramName);
  }
  return summaries;
}

/**
 * Format inference results as a table string
 *
 * @param result - The inference result
 * @returns Formatted table string
 */
export function formatSummaryTable(result: InferenceResult): string {
  const summaries = summarizeAll(result);
  const params = Object.keys(summaries);

  if (params.length === 0) {
    return 'No parameters in result';
  }

  // Header
  const lines: string[] = [
    `${'Parameter'.padEnd(15)} ${'Mean'.padStart(10)} ${'Std'.padStart(10)} ${'2.5%'.padStart(10)} ${'Median'.padStart(10)} ${'97.5%'.padStart(10)} ${'R-hat'.padStart(8)} ${'ESS'.padStart(8)}`,
    '-'.repeat(90),
  ];

  // Rows
  for (const param of params) {
    const s = summaries[param];
    lines.push(
      `${param.padEnd(15)} ${s.mean.toFixed(3).padStart(10)} ${s.std.toFixed(3).padStart(10)} ${s.q025.toFixed(3).padStart(10)} ${s.q50.toFixed(3).padStart(10)} ${s.q975.toFixed(3).padStart(10)} ${s.rhat.toFixed(3).padStart(8)} ${s.ess.toFixed(0).padStart(8)}`
    );
  }

  return lines.join('\n');
}

File: r-package/bayesiangpu/R/result.R
#' Inference Result
#'
#' An R6 class containing MCMC sampling results and diagnostics.
#'
#' @importFrom R6 R6Class
#' @importFrom tibble tibble
#' @export
InferenceResult <- R6::R6Class(
  "InferenceResult",
  public = list(
    #' @field samples Named list of sample vectors for each parameter
    samples = NULL,

    #' @field chains Named list of chain samples (list of vectors)
    chains = NULL,

    #' @field rhat Named list of R-hat values
    rhat = NULL,

    #' @field ess Named list of effective sample size values
    ess = NULL,

    #' @field divergences Number of divergent transitions
    divergences = NULL,

    #' @field step_size Final adapted step size
    step_size = NULL,

    #' @field num_samples Number of samples per chain
    num_samples = NULL,

    #' @field num_warmup Number of warmup iterations
    num_warmup = NULL,

    #' @field num_chains Number of chains
    num_chains = NULL,

    #' @field param_names Names of parameters
    param_names = NULL,

    #' @description Create a new InferenceResult from Rust output
    #' @param result_list List returned from run_nuts_sampling
    initialize = function(result_list) {
      self$samples <- result_list$samples
      self$chains <- result_list$chains
      self$rhat <- result_list$rhat
      self$ess <- result_list$ess
      self$divergences <- result_list$divergences
      self$step_size <- result_list$step_size
      self$num_samples <- result_list$num_samples
      self$num_warmup <- result_list$num_warmup
      self$num_chains <- result_list$num_chains
      self$param_names <- result_list$param_names
    },

    #' @description Get summary statistics for a parameter
    #' @param param_name Name of the parameter
    #' @return A list with mean, std, q025, q50, q975, rhat, ess
    summarize = function(param_name) {
      samples <- self$samples[[param_name]]
      list(
        mean = mean(samples),
        std = sd(samples),
        q025 = quantile(samples, 0.025, names = FALSE),
        q50 = quantile(samples, 0.5, names = FALSE),
        q975 = quantile(samples, 0.975, names = FALSE),
        rhat = self$rhat[[param_name]],
        ess = self$ess[[param_name]]
      )
    },

    #' @description Get summary for all parameters as a data frame
    #' @return A data frame with summary statistics
    summary = function() {
      rows <- lapply(self$param_names, function(name) {
        s <- self$summarize(name)
        data.frame(
          parameter = name,
          mean = s$mean,
          std = s$std,
          q2.5 = s$q025,
          q50 = s$q50,
          q97.5 = s$q975,
          rhat = s$rhat,
          ess = s$ess,
          stringsAsFactors = FALSE
        )
      })
      do.call(rbind, rows)
    },

    #' @description Format summary as a string
    #' @return Formatted summary string
    format_summary = function() {
      df <- self$summary()
      paste(capture.output(print(df, row.names = FALSE)), collapse = "\n")
    },

    #' @description Check for sampling warnings
    #' @return Vector of warning messages (empty if none)
    warnings = function() {
      warns <- character()

      # Check R-hat
      for (name in self$param_names) {
        if (self$rhat[[name]] > 1.01) {
          warns <- c(warns, sprintf("High R-hat for %s: %.3f (should be < 1.01)",
                                    name, self$rhat[[name]]))
        }
      }

      # Check ESS
      for (name in self$param_names) {
        if (self$ess[[name]] < 400) {
          warns <- c(warns, sprintf("Low ESS for %s: %.0f (should be > 400)",
                                    name, self$ess[[name]]))
        }
      }

      # Check divergences
      if (self$divergences > 0) {
        warns <- c(warns, sprintf("%d divergent transitions (should be 0)",
                                  self$divergences))
      }

      warns
    },

    #' @description Check if sampling converged
    #' @return TRUE if all R-hat < 1.01 and ESS > 400
    is_converged = function() {
      length(self$warnings()) == 0
    },

    #' @description Convert to a tibble (for ggplot2, dplyr, etc.)
    #' @return A tibble with samples
    as_tibble = function() {
      df <- as.data.frame(self$samples)
      tibble::as_tibble(df)
    },

    #' @description Print the result
    print = function() {
      cat("InferenceResult\n")
      cat(sprintf("  Chains: %d, Samples per chain: %d\n",
                  self$num_chains, self$num_samples))
      cat(sprintf("  Parameters: %s\n", paste(self$param_names, collapse = ", ")))
      cat(sprintf("  Divergences: %d\n", self$divergences))
      cat("\nSummary:\n")
      print(self$summary(), row.names = FALSE)

      warns <- self$warnings()
      if (length(warns) > 0) {
        cat("\nWarnings:\n")
        for (w in warns) {
          cat(sprintf("  - %s\n", w))
        }
      }

      invisible(self)
    }
  )
)

#' @export
as.data.frame.InferenceResult <- function(x, ...) {
  as.data.frame(x$samples)
}

#' @export
summary.InferenceResult <- function(object, ...) {
  object$summary()
}
